blocklib/timing/include/timing.hpp

```diff
@@ -0,0 +1,313 @@
+        condition->setAcceptEarly(true);
+        condition->setAcceptConflict(true);
+        condition->setAcceptDelayed(true);
+        condition->SigAction.connect(
+                [this](uint64_t id, uint64_t param, const saftlib::Time& deadline, const saftlib::Time& executed,
+                       uint16_t flags) {
+                    this->snoop_writer.publish(
+                            [id, param, &deadline, &executed, flags](std::span<Event> buffer) {
+                                buffer[0] = Timing::Event{deadline.getTAI(), id, param, flags, executed.getTAI()};
+                            }, 1);
+                });
+        condition->setActive(true);
+    }
+
+    void initialize() {
+        if (simulate) {
+            initialized = true;
+        } else {
+            try {
+                saftd = SAFTd_Proxy::create();
+                // get a specific device
+                std::map<std::string, std::string> devices = saftd->getDevices();
+                if (devices.empty()) {
+                    std::cerr << "No devices attached to saftd" << std::endl;
+                }
+                receiver = TimingReceiver_Proxy::create(devices.begin()->second);
+                sink = SoftwareActionSink_Proxy::create(receiver->NewSoftwareActionSink("gr_timing_example"));
+                updateSnoopFilter();
+                for (const auto & [i, output]: receiver->getOutputs() | std::views::enumerate ) {
+                    const auto &[name, port] = output;
+                    outputs.emplace_back(i, name, port);
+                }
+                initialized = true;
+            } catch (...) {}
+        }
+    }
+
+    void process() {
+        if (!initialized && !tried) {
+            tried = true;
+            initialize();
+        } else if (initialized & !simulate) {
+            const auto startTime = std::chrono::system_clock::now();
+            while(true) {
+                auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(startTime - std::chrono::system_clock::now() + std::chrono::milliseconds(5)).count();
+                if (duration > 0) {
+                    saftlib::wait_for_signal(static_cast<int>(std::clamp(duration, 50L, std::numeric_limits<int>::max()+0L)));
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    void injectEvent(const Event &ev, uint64_t time_offset) {
+        if (simulate && ((ev.id() & snoopMask) == (snoopID & snoopMask)) ) {
+            this->snoop_writer.publish(
+                    [ev, time_offset](std::span<Event> buffer) {
+                        buffer[0] = ev;
+                        buffer[0].time += time_offset;
+                        buffer[0].executed = buffer[0].time;
+                    }, 1);
+        } else if (!simulate) {
+            receiver->InjectEvent(ev.id(), ev.param(), saftlib::makeTimeTAI(ev.time + time_offset));
+        }
+    }
+
+    unsigned long getTAI() {
```

Comment:
needs better naming: TAI....

Reply:
Original name was forwarded from the saftlib api, but lost the currentTime part. Renamed to `currentTimeTAI()`. With camelCase acronyms kind of have to go in the back to keep things readable.